src/Razor/test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Cohost/CohostUriPresentationEndpointTest.cs

```diff
@@ -0,0 +1,269 @@
+using Microsoft.CodeAnalysis.Testing;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.VisualStudio.Razor.LanguageClient.Cohost;
+
+public class CohostUriPresentationEndpointTest(ITestOutputHelper testOutputHelper) : CohostEndpointTestBase(testOutputHelper)
+{
+    [Fact]
+    public async Task RandomFile()
+    {
+        await VerifyUriPresentationAsync(
+            input: """
+                This is a Razor document.
+
+                <div>
+                    [||]
+                </div>
+
+                The end.
+                """,
+            uris: [new("file:///C:/path/to/SomeRandomFile.txt")],
+            // In reality this would actual insert the full path, but the Html server does that for us, and we
+            // have other tests that validate that we insert what the Html server tells us
+            expected: null);
+    }
+
+    [Fact]
+    public async Task HtmlResponse_TranslatesVirtualDocumentUri()
+    {
+        var htmlTag = """<link src="file:///C:/path/to/site.css" rel="stylesheet" />""";
+
+        await VerifyUriPresentationAsync(
+            input: """
+                This is a Razor document.
+
+                <div>
+                    [||]
+                </div>
+
+                The end.
+                """,
+            uris: [new("file:///C:/path/to/site.css")],
+            htmlResponse: new WorkspaceEdit
+            {
+                DocumentChanges = new TextDocumentEdit[]
+                {
+                    new()
+                    {
+                        TextDocument = new()
+                        {
+                            Uri = new("file:///c:/users/example/src/SomeProject/File1.razor.g.html")
+                        },
+                        Edits = [new() { NewText = htmlTag}]
+                    }
+                }
+            },
+            expected: htmlTag);
+    }
+
+    [Fact]
+    public async Task Component()
+    {
+        await VerifyUriPresentationAsync(
+            input: """
+                This is a Razor document.
+
+                <div>
+                    [||]
+                </div>
+
+                The end.
+                """,
+            additionalFiles: [
+                // The source generator isn't hooked up to our test project, so we have to manually "compile" the razor file
+                (Path.Combine(TestProjectData.SomeProjectPath, "Component.cs"), """
+                    namespace SomeProject;
+
+                    public class Component : Microsoft.AspNetCore.Components.ComponentBase
+                    {
+                    }
+                    """),
+                // The above will make the component exist, but the .razor file needs to exist too for Uri presentation
+                (Path.Combine(TestProjectData.SomeProjectPath, "Component.razor"), """
+                    This doesn't matter
+                    """)
```

Comment:
Consider adding a helper method for all these tuples. All of them do a `Path.Combine(...)` with `TestProjectData.SomeProjectPath` as the first argument, so something like this might work:

```C#
private static (string filePath, string content) File(string fileName, string? content = null)
{
    content ??= """
    This doesn't matter
    """
    return (Path.Combine(TestProjectData.SomeProjectPath, fileName), content);
}
```

Then, each of these can be cleaned up like so:

```suggestion
                File("Component.cs", """
                    namespace SomeProject;

                    public class Component : Microsoft.AspNetCore.Components.ComponentBase
                    {
                    }
                    """),
                // The above will make the component exist, but the .razor file needs to exist too for Uri presentation
                File("Component.razor")
```

Reply:
Combined this with the FilePathUri idea, and did something slightly different.